src/renderer.rs

```diff
@@ -59,66 +60,42 @@ const RECT: [Vertex; 6] = [
-//
-// The struct might also be useless, and we could easily just
-// give out a SpriteSheet directly.
+/// A sprite sheet that lives on the GPU.
 #[derive(Clone, Debug)]
-pub struct SpriteSheetBuilder {
-    pub image: Vec<u8>,
-    pub image_dim: (Pixels, Pixels),
-    pub tile_dim: Option<(Pixels, Pixels)>,
-}
-
-/// A marker type for the unit pixels.
-pub type Pixels = usize;
-
-impl SpriteSheetBuilder {
-    pub fn new(width: Pixels, height: Pixels, image: Vec<u8>) -> Self {
-        Self {
-            image,
-            image_dim: (width, height),
-            tile_dim: None,
-        }
-    }
-
-    pub fn tile_size(mut self, tile_width: Pixels, tile_height: Pixels) -> Self {
-        self.tile_dim = Some((tile_width, tile_height));
-        self
-    }
-}
-
-/// A light reference to a SpriteSheet that lives on the GPU.
-#[derive(Clone, Copy, Debug)]
 pub struct SpriteSheet {
     id: usize,
-    image_dim: (Pixels, Pixels),
-    tile_dim: Option<(Pixels, Pixels)>,
+    image: Image,
+    tile_size: (Pixels, Pixels),
 }
 
 impl SpriteSheet {
     /// Returns the SpriteRegion of a tile given the specified tile sizes,
     /// starting from the top left.
     pub fn grid(&self, tx: usize, ty: usize) -> SpriteRegion {
-        if let Some(tile_dim) = self.tile_dim {
-            let xlo = ((tile_dim.0 * tx) as f32) / (SPRITE_SHEET_SIZE[0] as f32);
-            let ylo = ((tile_dim.1 * ty) as f32) / (SPRITE_SHEET_SIZE[1] as f32);
-            let w = (tile_dim.0 as f32) / (SPRITE_SHEET_SIZE[0] as f32);
-            let h = (tile_dim.1 as f32) / (SPRITE_SHEET_SIZE[1] as f32);
-            (
-                self.id as f32 / (SPRITE_SHEET_SIZE[2] as f32),
-                [xlo, ylo, xlo + w, ylo + h],
+        let xlo = ((self.tile_size.0 * tx) as f32) / (SPRITE_SHEET_SIZE[0] as f32);
+        let ylo = ((self.tile_size.1 * ty) as f32) / (SPRITE_SHEET_SIZE[1] as f32);
+        let w = (self.tile_size.0 as f32) / (SPRITE_SHEET_SIZE[0] as f32);
+        let h = (self.tile_size.1 as f32) / (SPRITE_SHEET_SIZE[1] as f32);
+        (
+            self.id as f32 / (SPRITE_SHEET_SIZE[2] as f32),
+            [xlo, ylo, xlo + w, ylo + h],
+        )
+    }
+
+    pub fn upload(&self, tex: &mut Tex) {
+        tex
+            .upload_part_raw(
+                GenMipmaps::No,
+                [0, 0, self.id as u32],
+                [self.image.width as u32, self.image.height as u32, 1],
+                &self.image.texture_data,
             )
-        } else {
-            // TODO(ed): We could potentially catch this as a type error
-            // using the SpriteSheet as an enum?
-            // TODO(ed):
-            // We could also find grid slices that are out of bounds,
-            // if we checked the image dimension.
-            panic!();
+            .unwrap();
+    }
+
+    pub fn reload(&mut self, tex: &mut Tex) {
+        if self.image.reload() {
+            self.upload(tex);
         }
     }
 }
```

Comment:
Should this live here? Or should it live in asset.rs?

Reply:
It should *probably* live in asset.rs but I think it's part of a larger refactor. I'd rather keep it here for the time being and do a larger refactor later.